rose/learner.py

```diff
@@ -346,3 +346,159 @@ def _parallel_active_learn():
+    training tasks, but distinct active learning task.
+    """
+    def __init__(self, engine: ResourceEngine) -> None:
+        ''' 
+        Initialize the ParallelActiveLearningAlgoSelector object.
+
+        Args:
+            engine: The ResourceEngine object that manages the resources and
+            tasks submission to HPC resources during the active learning loop.
+        '''
+        super().__init__(engine, register_and_submit=False)
+        self.active_learn_functions: Dict[str, Dict] = {}
+
+        # A dictionary to store stats for each active learning pipeline
+        # e.g. self._pipeline_stats['algo_1'] = {'iterations': 5, 'last_result': 0.01}
+        self._pipeline_stats: Dict[str, Dict] = {}
+        self.best_pipeline_name = None
+        self.best_pipeline_stats = None
+
+    def _check_stop_criterion_and_return_value(self, stop_task_result):
+
+        try:
+            metric_value = eval(stop_task_result)
+        except Exception as e:
+            raise Exception(f"Failed to obtain a numerical value from criterion task: {e}")
+
+        # check if the metric value is a number
+        if isinstance(metric_value, float) or isinstance(metric_value, int):
+            operator = self.criterion_function['operator']
+            threshold = self.criterion_function['threshold']
+            metric_name = self.criterion_function['metric_name']
+
+            if self.compare_metric(metric_name, metric_value, threshold, operator):
+                print(f'stop criterion metric: {metric_name} is met with value of: {metric_value}'\
+                      '. Breaking the active learning loop')
+                return True, metric_value
+            else:
+                print(f'stop criterion metric: {metric_name} is not met yet ({metric_value}).')
+                return False, metric_value
+        else:
+            raise TypeError(f'Stop criterion task must produce a numerical value, got {type(metric_value)} instead')
+
+    def active_learn_task(self, name: str):
+        """
+        A decorator that registers an active learning task under the given name.
+        
+        Usage:
+        
+        @algo_selector.active_learn_task(name='algo_1')
+        def active_learn_1(*args):
+            ...
+        
+        @algo_selector.active_learn_task(name='algo_2')
+        def active_learn_2(*args):
+            ...
+        """
+        def decorator(func: Callable):
+            @wraps(func)
+            def wrapper(*args, **kwargs):
+                self.active_learn_functions[name] = {
+                    'func': func,
+                    'args': args,
+                    'kwargs': kwargs
+                }
+                if self.register_and_submit:
+                    return self._register_task(self.active_learn_functions[name])
+            return wrapper
+        return decorator
+
+    def teach(self, max_iter:int = 0, skip_pre_loop:bool = False):
```

Comment:
I think calling this `teach_and_select` makes more sense as there is a selection process in addition to the `teaching`